### PR TITLE
feat: Add Apigee Envgroup direct controller

### PR DESCRIPTION
**Pull request by Camila-B**
### Change description

Defines the Apigee Envgroup API

### Tests you have done

- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [x] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: apis/apigee/v1alpha1/types.generated.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+// +kcc:proto=mockgcp.cloud.apigee.v1.GoogleCloudApigeeV1EnvironmentGroup
+type GoogleCloudApigeeV1EnvironmentGroup struct {
```
Comment (jasonvigil) [inline apis/apigee/v1alpha1/types.generated.go:18]: You may need to comment this struct out (including above comment marker), to avoid confusing mapper generator. There should only be two structs tagged `+kcc:proto=mockgcp.cloud.apigee.v1.GoogleCloudApigeeV1EnvironmentGroup` (spec and observed state)

File: apis/apigee/v1alpha1/googlecloudapigeev1environmentgroup_types.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
+// Copyright 2024 Google LLC
```
Comment (jasonvigil) [inline apis/apigee/v1alpha1/googlecloudapigeev1environmentgroup_types.go:1]: For consistency with other API file names, would suggest renaming this (and others) to remove the `googlecloudapigeev1` prefix.

**Comment by jasonvigil** [comment]
Seems to be a failure in presubmit validations, looks like some of the generated files need to be regenerated:
```
ERROR: Manifests must be regenerated. Please run 'make ready-pr' or 'make manifests' and update your PR.
```

**Comment by jasonvigil** [comment]
Also, looks like you fixed one of the issues in `tests/apichecks/testdata/exceptions/acronyms.txt`, and we can remove the exception now! See unit test job output: https://github.com/GoogleCloudPlatform/k8s-config-connector/actions/runs/12203794855/job/34047550729?pr=3332

File: apis/apigee/v1alpha1/types.generated.go (hunk 1 above)
Comment (Camila-B) [reply apis/apigee/v1alpha1/types.generated.go]: Done

File: apis/apigee/v1alpha1/googlecloudapigeev1environmentgroup_types.go (hunk 2 above)
Comment (Camila-B) [reply apis/apigee/v1alpha1/googlecloudapigeev1environmentgroup_types.go]: Done

File: apis/apigee/v1alpha1/environmentgroup_identity.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+// 			Location:  location,
+// 		},
+// 		id: resourceID,
+// 	}, nil
+// }
+
+func ParseGoogleCloudApigeeV1EnvironmentGroupExternal(external string) (parent *GoogleCloudApigeeV1EnvironmentGroupParent, resourceID string, err error) {
+	tokens := strings.Split(external, "/")
+	if len(tokens) != 6 || tokens[0] != "projects" || tokens[2] != "locations" || tokens[4] != "googlecloudapigeev1environmentgroups" {
+		return nil, "", fmt.Errorf("format of ApigeeEnvgroup external=%q was not known (use projects/<projectId>/locations/<location>/googlecloudapigeev1environmentgroups/<googlecloudapigeev1environmentgroupID>)", external)
```
Comment (jasonvigil) [inline apis/apigee/v1alpha1/environmentgroup_identity.go]: After https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/3398, this should be updated to use `{{` and `}}` instead of `<` and `>`:
```
return nil, "", fmt.Errorf("format of ApigeeEnvgroup external=%q was not known (use projects/{{projectId}}/locations/{{location}}/googlecloudapigeev1environmentgroups/{{googlecloudapigeev1environmentgroupID}})", external)
```
This is more for consistency with the API (which actually needs to be updated so our g3docs render properly), but good to do.

File: apis/apigee/v1alpha1/environmentgroup_reference.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ refsv1beta1.ExternalNormalizer = &GoogleCloudApigeeV1EnvironmentGroupRef{}
+
+// GoogleCloudApigeeV1EnvironmentGroupRef defines the resource reference to ApigeeEnvgroup, which "External" field
+// holds the GCP identifier for the KRM object.
+type GoogleCloudApigeeV1EnvironmentGroupRef struct {
+	// A reference to an externally managed ApigeeEnvgroup resource.
+	// Should be in the format "projects/<projectID>/locations/<location>/googlecloudapigeev1environmentgroups/<googlecloudapigeev1environmentgroupID>".
```
Comment (jasonvigil) [inline apis/apigee/v1alpha1/environmentgroup_reference.go:35]: After https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/3398, this needs to be updated:
```
// Should be in the format "projects/{{projectID}}/locations/{{location}}/googlecloudapigeev1environmentgroups/{{googlecloudapigeev1environmentgroupID}}".
```
And, re-run `make manifests` / `make resource-docs`. This is so our g3docs render properly.

File: apis/apigee/v1alpha1/environmentgroup_types.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+type Parent struct {
+	// +required
+	OrganizationRef *refv1alpha1.OrganizationRef `json:"organizationRef"`
+}
+
+// ApigeeEnvgroupSpec defines the desired state of ApigeeEnvgroup
+// +kcc:proto=mockgcp.cloud.apigee.v1.GoogleCloudApigeeV1EnvironmentGroup
+type ApigeeEnvgroupSpec struct {
+	Parent `json:",inline"`
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="ResourceID field is immutable"
```
Comment (jasonvigil) [inline apis/apigee/v1alpha1/environmentgroup_types.go:34]: This needs to be removed, template is getting updated in https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/3434. Also remove the `// Immutable.` comment. We are allowing changes to this field, so the user can set from empty string -> the actual resource ID, if they want to. Otherwise, it is immutable.

You will need to run `make manifests` (and maybe `make resource-docs`) after this change also.

File: apis/apigee/v1alpha1/environmentgroup_types.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	CreatedAt *int64 `json:"createdAt,omitempty"`
+	// Output only. The time at which the environment group was last updated as milliseconds since epoch.
+	LastModifiedAt *int64 `json:"lastModifiedAt,omitempty"`
+	// Output only. State of the environment group. Values other than ACTIVE means the resource is not ready to use.
+	State *string `json:"state,omitempty"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// TODO(user): make sure the pluralizaiton below is correct
```
Comment (jasonvigil) [inline apis/apigee/v1alpha1/environmentgroup_types.go:74]: Nit: remove todo comment

File: dev/tools/controllerbuilder/pkg/commands/generatetypes/generatetypescommand.go [hunk 7]
```diff
@@ -173,6 +173,10 @@ func RunGenerateCRD(ctx context.Context, o *GenerateCRDOptions) error {
 			return err
 		}
 
+		if err := typeGenerator.WriteVisitedMessages(); err != nil {
```
Comment (jasonvigil) [inline dev/tools/controllerbuilder/pkg/commands/generatetypes/generatetypescommand.go:176]: I think we should remove the changes for this file. It's not necessarily a long-term fix (more of a hack to unblock for generation temporarily).

File: apis/refs/v1alpha1/organizationref.go [hunk 8]
```diff
@@ -27,6 +39,61 @@ type Organization struct {
...
+	organization := &unstructured.Unstructured{}
+	organization.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "apigee.cnrm.cloud.google.com",
+		Version: "v1alpha1",
+		Kind:    "ApigeeOrganization",
+	})
+
+	if err := reader.Get(ctx, key, organization); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, fmt.Errorf("referenced Organization %v not found", key)
```
Comment (jasonvigil) [inline apis/refs/v1alpha1/organizationref.go:87]: Here, we should return a `k8s.NewReferenceNotFoundError`, so that we will immediately requeue once the reference is found. Example: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/26f6f8a355f2a23933e63f48610f915c1022dddd/apis/workstations/v1beta1/cluster_reference.go#L70C15-L70C44

File: pkg/controller/direct/apigee/envgroup_controller.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+// Update updates the resource in GCP based on `spec` and update the Config Connector object `status` based on the GCP response.
+func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperation) error {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("patching ApigeeEnvgroup", a.fullyQualifiedName())
+
+	updateMask := fieldmaskpb.FieldMask{}
+
+	if !reflect.DeepEqual(a.desired.Hostnames, a.actual.Hostnames) {
```
[the remark this answers is not in the document]
Comment (Camila-B) [reply pkg/controller/direct/apigee/envgroup_controller.go]: Done

File: apis/refs/v1alpha1/organizationref.go (hunk 8 above)
Comment (Camila-B) [reply apis/refs/v1alpha1/organizationref.go]: Done

File: pkg/controller/direct/apigee/envgroup_mappings.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+package apigee
+
+import (
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/apigee/v1alpha1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	api "google.golang.org/api/apigee/v1"
+)
+
+func ApigeeEnvgroupObservedState_FromApi(mapCtx *direct.MapContext, in *api.GoogleCloudApigeeV1EnvironmentGroup) *krm.ApigeeEnvgroupObservedState {
```
[the remark this answers is not in the document]
Comment (Camila-B) [reply pkg/controller/direct/apigee/envgroup_mappings.go:23]: Done

File: dev/tools/controllerbuilder/pkg/commands/generatetypes/generatetypescommand.go (hunk 7 above)
Comment (Camila-B) [reply dev/tools/controllerbuilder/pkg/commands/generatetypes/generatetypescommand.go]: Done

File: apis/apigee/v1alpha1/environmentgroup_types.go (hunk 6 above)
Comment (Camila-B) [reply apis/apigee/v1alpha1/environmentgroup_types.go]: Done

File: apis/apigee/v1alpha1/environmentgroup_types.go (hunk 5 above)
Comment (Camila-B) [reply apis/apigee/v1alpha1/environmentgroup_types.go]: Done

File: apis/apigee/v1alpha1/environmentgroup_reference.go (hunk 4 above)
Comment (Camila-B) [reply apis/apigee/v1alpha1/environmentgroup_reference.go]: Done. I also updated it to reflect the actual format used `organizations/{{organization}}/envgroups/{{envgroup}}`

File: pkg/controller/direct/apigee/envgroup_controller.go [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+	updateMask := fieldmaskpb.FieldMask{}
+
+	// Sorts the Hostname lists so that the comparison is deterministic
+	if !reflect.DeepEqual(asSortedCopy(a.desired.Hostnames), asSortedCopy(a.actual.Hostnames)) {
+		log.V(2).Info("change detected: hostnames")
+		updateMask.Paths = append(updateMask.Paths, "hostnames")
+	}
+
+	if len(updateMask.Paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id)
```
Comment (jasonvigil) [inline pkg/controller/direct/apigee/envgroup_controller.go:172]: Here, we need to update status also. This is a recent change to the controller template. When adopting existing resources, it might be the case that no fields actually change, but we still need to update the status to print output only fields. Example: https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/680a1271c0a6ca2b16ceb972820de8d923a0cd20/pkg/controller/direct/workstations/workstation_controller.go#L211-L217

Comment (Camila-B) [reply pkg/controller/direct/apigee/envgroup_controller.go:172]: Done

**Review by yuwenma — COMMENTED** [review]
/hold

**Review by yuwenma — COMMENTED** [review]
/hold

File: apis/apigee/v1alpha1/environmentgroup_identity.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+	"strings"
+
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1alpha1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GoogleCloudApigeeV1EnvironmentGroupIdentity defines the resource reference to ApigeeEnvgroup, which "External" field
+// holds the GCP identifier for the KRM object.
+type GoogleCloudApigeeV1EnvironmentGroupIdentity struct {
```
Comment (yuwenma) [inline apis/apigee/v1alpha1/environmentgroup_identity.go:29]: nit: can we shorten this to `EnvironmentGroupIdentity`? Here and other places

File: apis/refs/v1alpha1/helper.go
Comment (yuwenma) [inline apis/refs/v1alpha1/helper.go:1]: Suggest using the same functions under refs/v1beta1/

File: apis/refs/v1alpha1/organizationref.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+type Organization struct {
+	OrganizationName string
+}
+
+func (s *Organization) FullyQualifiedName() string {
+	return "organizations/" + s.OrganizationName
+}
+func ResolveOrganization(ctx context.Context, reader client.Reader, src client.Object, ref *OrganizationRef) (*Organization, error) {
```
Comment (yuwenma) [inline apis/refs/v1alpha1/organizationref.go:46]: Can we implement the `NormalizeExternal` instead of ResolveOrganization?   Resolve{{Kind}} is the old format and have  problems when integrating with its referred code

File: pkg/controller/direct/apigee/envgroup_controller.go [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+	api "google.golang.org/api/apigee/v1"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	serviceDomain = "//apigee.googleapis.com"
```
Comment (yuwenma) [inline pkg/controller/direct/apigee/envgroup_controller.go:41]: nit: we no longer need the serviceDomain. The externalRef should be the relative URL without version and domain

File: pkg/controller/direct/apigee/envgroup_controller.go [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	id, err := krm.NewGoogleCloudApigeeV1EnvironmentGroupIdentity(ctx, reader, obj)
+	if err != nil {
+		return nil, err
+	}
+
+	mapCtx := &direct.MapContext{}
+	desired := ApigeeEnvgroup_ToApi(mapCtx, obj)
```
Comment (yuwenma) [inline pkg/controller/direct/apigee/envgroup_controller.go:81]: Suggest keeping the KRM object format. It contains more information than proto object (ie.g. resource reference, spec/status diff)

File: pkg/controller/direct/apigee/envgroup_controller.go [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+	desired              *api.GoogleCloudApigeeV1EnvironmentGroup
+	actual               *api.GoogleCloudApigeeV1EnvironmentGroup
+	apigeeEnvgroupClient *apigeeEnvgroupClient
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+// Find retrieves the GCP resource.
+// Return true means the object is found. This triggers Adapter `Update` call.
+// Return true means the object is not found. This triggers Adapter `Create` call.
```
Comment (yuwenma) [inline pkg/controller/direct/apigee/envgroup_controller.go:109]: nit: typo `Return false`

File: pkg/controller/direct/apigee/envgroup_controller.go [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+func (m *modelApigeeEnvgroup) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id                   *krm.GoogleCloudApigeeV1EnvironmentGroupIdentity
+	desired              *api.GoogleCloudApigeeV1EnvironmentGroup
+	actual               *api.GoogleCloudApigeeV1EnvironmentGroup
+	apigeeEnvgroupClient *apigeeEnvgroupClient
```
Comment (yuwenma) [inline pkg/controller/direct/apigee/envgroup_controller.go:102]: Suggest splitting apigeeEnvgroupClient to gcp client and operator client.  It is more readable in the following code (create, waitforOpt) because for KCC controller they represents two different kinds of services and HTTP calls.

File: pkg/controller/direct/apigee/envgroup_controller.go [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+			return fmt.Errorf("getting operation status of %q: %w", op.Name, err)
+		}
+		if current.Done {
+			if current.Error != nil {
+				return fmt.Errorf("operation %q completed with error: %v", op.Name, current.Error)
+			} else {
+				return nil
+			}
+		}
+		time.Sleep(2 * time.Second)
```
Comment (yuwenma) [inline pkg/controller/direct/apigee/envgroup_controller.go:264]: How long does this update normally takes?

File: pkg/controller/direct/apigee/envgroup_mappings.go [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+		return nil
+	}
+	out := &api.GoogleCloudApigeeV1EnvironmentGroup{}
+	out.Hostnames = in.Spec.Hostnames
+
+	if in.Spec.ResourceID != nil {
+		out.Name = *in.Spec.ResourceID
+	} else {
+		out.Name = in.Name
+	}
```
Comment (yuwenma) [inline pkg/controller/direct/apigee/envgroup_mappings.go:68]: We should not need this function but use the adaptor.id to get the `Name`

File: pkg/test/resourcefixture/testdata/basic/apigee/v1alpha1/apigeeenvgroup/apigeeenvgroup-basic/_http.log [hunk 20]
```diff
@@ -56,21 +55,53 @@ X-Xss-Protection: 0
   "name": "organizations/${projectId}/operations/${operationID}",
   "response": {
     "@type": "type.googleapis.com/google.cloud.apigee.v1.GoogleCloudApigeeV1EnvironmentGroup",
-    "createdAt": "2024-04-01T12:34:56.123456Z",
```
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/apigee/v1alpha1/apigeeenvgroup/apigeeenvgroup-basic/_http.log:59]: Can we improve the mock test to include this field (and `lastModifiedAt` below)? The KCC API uses *int64 to hold these two fields, we want to make sure it is correctly mapped in status.

File: pkg/test/resourcefixture/testdata/basic/apigee/v1alpha1/apigeeenvgroup/apigeeenvgroup-basic/_generated_object_apigeeenvgroup-basic.golden.yaml [hunk 21]
```diff
@@ -15,13 +14,20 @@ metadata:
...
   conditions:
   - lastTransitionTime: "1970-01-01T00:00:00Z"
     message: The resource is up to date
     reason: UpToDate
     status: "True"
     type: Ready
+  externalRef: organizations/${organizationID}/envgroups/apigeeenvgroup-${uniqueId}
   observedGeneration: 2
+  observedState:
+    createdAt: "1970-01-01T00:00:00Z"
```
Comment (yuwenma) [inline pkg/test/resourcefixture/testdata/basic/apigee/v1alpha1/apigeeenvgroup/apigeeenvgroup-basic/_generated_object_apigeeenvgroup-basic.golden.yaml:30]: I'm confused. This is *int64. How does this map to a string in golden log?

File: tests/apichecks/testdata/exceptions/acronyms.txt [hunk 22]
```diff
@@ -19,7 +19,6 @@
 [acronyms] crd=apigeeaddonsconfigs.apigee.cnrm.cloud.google.com version=v1alpha1: field ".spec.addonsConfig.advancedApiOpsConfig.enabled" should be ".spec.addonsConfig.advancedAPIOpsConfig.enabled"
 [acronyms] crd=apigeeendpointattachments.apigee.cnrm.cloud.google.com version=v1alpha1: field ".spec.orgId" should be ".spec.orgID"
 [acronyms] crd=apigeeenvgroupattachments.apigee.cnrm.cloud.google.com version=v1alpha1: field ".spec.envgroupId" should be ".spec.envgroupID"
-[acronyms] crd=apigeeenvgroups.apigee.cnrm.cloud.google.com version=v1alpha1: field ".spec.orgId" should be ".spec.orgID"
```
Comment (yuwenma) [inline tests/apichecks/testdata/exceptions/acronyms.txt:22]: Nice!

**Review by yuwenma — COMMENTED** [review]
I have a few questions related to the LRO and returned value format. Can we upload the real GCP log to show the git-diff?

File: apis/apigee/v1alpha1/environmentgroup_identity.go [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+func (p *GoogleCloudApigeeV1EnvironmentGroupParent) String() string {
+	return "organizations/" + p.Organization
+}
+
+// New builds a GoogleCloudApigeeV1EnvironmentGroupIdentity from the Config Connector GoogleCloudApigeeV1EnvironmentGroup object.
+func NewGoogleCloudApigeeV1EnvironmentGroupIdentity(ctx context.Context, reader client.Reader, obj *ApigeeEnvgroup) (*GoogleCloudApigeeV1EnvironmentGroupIdentity, error) {
+	// Get Parent
+	orgRef, err := refs.ResolveOrganization(ctx, reader, obj, obj.Spec.Parent.OrganizationRef)
```
[the remark this answers is not in the document]
Comment (jasonvigil) [reply apis/apigee/v1alpha1/environmentgroup_identity.go:57]: I am not sure that it will be possible to avoid resolving the reference here, because the organization is the parent for this resource. Even in `NormalizeExternal`, we resolve the resource, if it is not specified in external format.

File: apis/refs/v1alpha1/helper.go
Comment (Camila-B) [reply apis/refs/v1alpha1/helper.go:1]: Done

File: apis/apigee/v1alpha1/environmentgroup_identity.go (hunk 12 above)
Comment (Camila-B) [reply apis/apigee/v1alpha1/environmentgroup_identity.go]: Done. I also shortened `GoogleCloudApigeeV1EnvironmentGroupRef` to `EnvironmentGroupRef`

File: apis/apigee/v1alpha1/environmentgroup_identity.go [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+
+// GoogleCloudApigeeV1EnvironmentGroupIdentity defines the resource reference to ApigeeEnvgroup, which "External" field
+// holds the GCP identifier for the KRM object.
+type GoogleCloudApigeeV1EnvironmentGroupIdentity struct {
+	parent *GoogleCloudApigeeV1EnvironmentGroupParent
+	id     string
+}
+
+func (i *GoogleCloudApigeeV1EnvironmentGroupIdentity) String() string {
+	return fmt.Sprintf("%s/envgroups/%s", i.parent.String(), i.id)
```
[the remark this answers is not in the document]
Comment (Camila-B) [reply apis/apigee/v1alpha1/environmentgroup_identity.go]: Done

File: pkg/controller/direct/apigee/envgroup_mappings.go (hunk 19 above)
Comment (Camila-B) [reply pkg/controller/direct/apigee/envgroup_mappings.go]: Got it. I removed this block of code

File: apis/refs/v1alpha1/organizationref.go (hunk 13 above)
Comment (Camila-B) [reply apis/refs/v1alpha1/organizationref.go]: Done

File: pkg/controller/direct/apigee/envgroup_controller.go (hunk 14 above)
Comment (Camila-B) [reply pkg/controller/direct/apigee/envgroup_controller.go]: Done

File: pkg/controller/direct/apigee/envgroup_controller.go (hunk 16 above)
Comment (Camila-B) [reply pkg/controller/direct/apigee/envgroup_controller.go]: Done

File: pkg/test/resourcefixture/testdata/basic/apigee/v1alpha1/apigeeenvgroup/apigeeenvgroup-basic/_http.log (hunk 20 above)
Comment (Camila-B) [reply pkg/test/resourcefixture/testdata/basic/apigee/v1alpha1/apigeeenvgroup/apigeeenvgroup-basic/_http.log:59]: I ran `hack/record-gcp` and `hack/compare-mock` again so both fields are part of the envgroup response

File: pkg/controller/direct/apigee/envgroup_controller.go (hunk 17 above)
Comment (Camila-B) [reply pkg/controller/direct/apigee/envgroup_controller.go]: Done

File: pkg/controller/direct/apigee/envgroup_controller.go (hunk 15 above)
Comment (Camila-B) [reply pkg/controller/direct/apigee/envgroup_controller.go]: Done. I changed `desired` to use the KRM object format

File: pkg/controller/direct/apigee/envgroup_controller.go (hunk 18 above)
Comment (Camila-B) [reply pkg/controller/direct/apigee/envgroup_controller.go:270]: From testing, it seems to be close to instantaneous (< 1 second)

**Comment by acpana** [comment]
@yuwenma could we merge this in so I can build on top of it too 😆 ?

**Comment by Camila-B** [comment]
Is there anything else that I need to change before this PR can be approved @yuwenma?

**Comment by yuwenma** [comment]
from the validation err, it seems that this PR is based from a very stale git-commit (which makes me concerns about if it contains our recent changes for API and controllers). @Camila-B Could you rebase the master head and upload the change in a new commit so we can easily verify that?
